src/encoding/bencode.rs

```diff
@@ -1,136 +1,125 @@
+encode_int!(isize); encode_int!(usize);
 
-impl From<isize> for Int {
-    fn from(content: isize) -> Int {
-        Int { content }
-    }
-}
 
 /*
 **  Lists are encoded as follows: l<bencoded values>e
 **  The initial l and trailing e are beginning and ending delimiters.
 **  Lists may contain any bencoded type, including integers, strings, dictionaries, and even lists within other lists.
 */
-
-pub struct List {
-    content: Vec<ListEntry>
-}
-
-impl Literal for List {
-    fn stringify(&self) -> Result<String, Error> {
-        let mut ascii = String::new();
-        for i in self.content.iter() {
-            let result = i.stringify()?;
-            ascii.push_str(result.as_str());
+impl Encodable for Vec<Box<dyn Encodable>> {
+    fn encode(&self) -> Result<String, Error> {
+        let mut bstring: Vec<char> = Vec::new();
+        
+        for item in self {
+            bstring.extend(item.encode()?.chars())
         }
-        Ok(format!("l{}e", ascii))
-    }
-}
 
-impl From<Vec<ListEntry>> for List {
-    fn from(content: Vec<ListEntry>) -> Self {
-        List { content }
+        Ok(format!("l{}e", bstring.iter().collect::<String>()))
     }
 }
 
+
 /*
 **  Dictionaries are encoded as follows: d<bencoded string><bencoded element>e
 **  The initial d and trailing e are the beginning and ending delimiters.
 **  The values may be integers, strings, lists, and other dictionaries.
 */
-
-pub struct Dict {
-    content: Vec<DictEntry>
+// Dictionary encoding is seperate from encoding a regular vector as each entry needs to be sorted in
+// lexicogrpahical order of the key as per specification.
+pub type Dictionary = Vec<DictionaryEntry>;
+pub struct DictionaryEntry {
+    key: String,
+    elem: Box<dyn Encodable>
 }
 
-impl Literal for Dict {
-    fn stringify(&self) -> Result<String, Error> {
-        let mut ascii = String::new();
-        for (key, value) in self.content.iter() {
-            let key = key.stringify()?;
-            let value = value.stringify()?;
-            ascii.push_str(key.as_str());
-            ascii.push_str(value.as_str());
+impl DictionaryEntry {
+    /// Create a new dictionary entry from a key and an encodable element
+    pub fn new<T: Encodable+'static>(key: &str, elem: T) -> Self {
+        Self { 
+            key: key.to_string(),
+            elem: Box::new(elem)
         }
-        Ok(format!("d{}e", ascii))
     }
 }
 
-impl From<Vec<DictEntry>> for Dict {
-    fn from(content: Vec<DictEntry>) -> Self {
-        Dict { content }
+impl Encodable for DictionaryEntry {
+    fn encode(&self) -> Result<String, Error> {
+        Ok(
+            format!("{}{}", self.key.encode()?, self.elem.encode()?)
+        )
+    }
+}
+
+impl Encodable for Vec<DictionaryEntry> {
+    fn encode(&self) -> Result<String, Error> {
+        let mut entries = Vec::new();
+        for item in self {
+            entries.push(item);
+        }
+        entries.sort_by(|a, b| a.key.cmp(&b.key));
```

Comment:
- `|a, b|` Never seen this syntax pls explain
- What does `cmp` do, and what is its purpose here

Reply:
`| |` is the syntax for closures. Anonymous functions.
Documentation for closures: https://doc.rust-lang.org/book/ch13-01-closures.html

Documentation for `cmp`: https://doc.rust-lang.org/std/cmp/index.html

Reply:
Ok, basically like arrow functions in JS, good to know 🚀